# Project 3: Reddit Web Scraping File

### Import Libraries

In [1]:
import requests
import pandas as pd
import time
import random
pd.set_option('display.max_columns', None)

### Identify the 2 Subreddits for analysis

In [2]:
# Declare the URLs for democrats and republican data

dem_url = 'https://www.reddit.com/r/democrats.json'
rep_url = 'https://www.reddit.com/r/republican.json'

### Create custom function to perform Reddit scrape

<font color=blue> A custom function is created to scrape based on the :
    <ol>
<li>Web address provided</li> 
<li>Number of pages intended for scraping</li>
    </ol>
    
A random interval of between 2-30 seconds of rest will be applied to prevent API detecting the script as malicious.
</font>

In [3]:
# Create custom function where you input URL and number of scrapes

posts = []

def web_scraper(url,scrape_count):
    after = None
    count = int(scrape_count) 
    
    for i in range(count):
        if after == None:
            current_url = url
        else:
            current_url = url + '?after=' + after + '&limit=50'
        print(current_url)
        res = requests.get(current_url, headers={'User-agent': 'lphong'})
            
        if res.status_code != 200:
            print('Status error', res.status_code)
            break  
                
        current_dict = res.json()
        current_posts = [p['data'] for p in current_dict['data']['children']]
        posts.extend(current_posts)
        after = current_dict['data']['after']
    
        # generate a random sleep duration to look more 'natural'
        sleep_duration = random.randint(2,30)
        print(sleep_duration)
        time.sleep(sleep_duration)

### <font color=red>Scrape Democrats Subreddit</font>

In [4]:
# Feed the subreddit URL and scrape for 20 x 50 posts (Reddit's 1,000 post limit)

web_scraper(dem_url,'21')

https://www.reddit.com/r/democrats.json
21
https://www.reddit.com/r/democrats.json?after=t3_hetkt8&limit=50
28
https://www.reddit.com/r/democrats.json?after=t3_heq3fc&limit=50
18
https://www.reddit.com/r/democrats.json?after=t3_hddclw&limit=50
19
https://www.reddit.com/r/democrats.json?after=t3_hcud4a&limit=50
3
https://www.reddit.com/r/democrats.json?after=t3_hcos8m&limit=50
17
https://www.reddit.com/r/democrats.json?after=t3_hboveu&limit=50
2
https://www.reddit.com/r/democrats.json?after=t3_hbicgh&limit=50
15
https://www.reddit.com/r/democrats.json?after=t3_h9qke4&limit=50
16
https://www.reddit.com/r/democrats.json?after=t3_h9e6ac&limit=50
18
https://www.reddit.com/r/democrats.json?after=t3_h8k91q&limit=50
25
https://www.reddit.com/r/democrats.json?after=t3_h7ktwz&limit=50
9
https://www.reddit.com/r/democrats.json?after=t3_h7i7x3&limit=50
10
https://www.reddit.com/r/democrats.json?after=t3_h0i2o9&limit=50
7
https://www.reddit.com/r/democrats.json?after=t3_gzwx3u&limit=50
7
https://ww

In [5]:
# Load information into a dataframe and reset post for next pull

dem_data = pd.DataFrame(posts)
posts = []

In [6]:
# Check the number of posts scraped and size up data

dem_data.shape

(990, 112)

In [7]:
# Inspect the dataframe to identify the relevant fields for further extraction and analysis

dem_data.head()

approved_at_utc  subreddit selftext author_fullname  saved mod_reason_title  \
0            None  democrats                 t2_tkz7y  False             None   
1            None  democrats              t2_713c2cph  False             None   
2            None  democrats              t2_35r4k0nu  False             None   
3            None  democrats                 t2_y4w5p  False             None   
4            None  democrats                  t2_1woh  False             None   

   gilded  clicked                                              title  \
0       0    False  Hey All! I made a documentary about Joe Biden....   
1       0    False      It’s Time for a Blue Wave to Restore America.   
2       0    False  In Scathing Letter, More Than 80-Percent of Fa...   
3       0    False  Mark Cuban endorses Biden on Hannity: He 'actu...   
4       0    False  Biden Surge and Trump Failures Expand Electora...   

  link_flair_richtext subreddit_name_prefixed  hidden  pwls  \
0                  []             r/democrats   False     6   
1                  []             r/democrats   False     6   
2                  []             r/democrats   False     6   
3                  []             r/democrats   False     6   
4                  []             r/democrats   False     6   

  link_flair_css_class  downs  thumbnail_height top_awarded_type  hide_score  \
0                 None      0             105.0             None       False   
1                 None      0             140.0             None       False   
2                 None      0              73.0             None       False   
3                 blue      0              78.0             None       False   
4                 None      0             104.0             None        True   

        name  quarantine link_flair_text_color  upvote_ratio  \
0  t3_hdge6q       False                  dark          0.71   
1  t3_hff6bt       False                  dark          0.93   
2  t3_hf7pni       False                  dark          0.97   
3  t3_hfgpv1       False                  dark          0.97   
4  t3_hfmomx       False                  dark          1.00   

  author_flair_background_color subreddit_type   ups  total_awards_received  \
0                          None         public    25                      1   
1                          None         public   837                      0   
2                          None         public  1354                      0   
3                          None         public   101                      0   
4                          None         public     8                      0   

                                         media_embed  thumbnail_width  \
0  {'content': '&lt;iframe width="600" height="33...            140.0   
1                                                 {}            140.0   
2                                                 {}            140.0   
3                                                 {}            140.0   
4                                                 {}            140.0   

  author_flair_template_id  is_original_content user_reports  \
0                     None                False           []   
1                     None                False           []   
2                     None                False           []   
3                     None                False           []   
4                     None                False           []   

                                        secure_media  is_reddit_media_domain  \
0  {'type': 'youtube.com', 'oembed': {'provider_u...                   False   
1                                               None                   False   
2                                               None                   False   
3                                               None                   False   
4                                               None                   False   

   is_meta category                                

### Save and export raw uncleaned file before treatment

In [8]:
dem_data.to_csv('../democrat_uncleaned.csv', index=False, encoding='utf-8-sig')

In [9]:
# Load the uncleaned file into a new dataframe for cleaning

democrats = pd.read_csv('../democrat_uncleaned.csv')

In [10]:
# The ones with discussion text for analysis are mainly title, selftext and comments. (comments excluded due to volume)
# "Distinguished" will be pulled as well to identify moderator posts and see if they should be kept

cols_to_keep = ['subreddit','title','selftext','distinguished']
democrats = democrats.loc[:, cols_to_keep]

In [11]:
# Verify that all data was carried over

democrats.shape

(990, 4)

In [12]:
democrats.head()

subreddit                                              title selftext  \
0  democrats  Hey All! I made a documentary about Joe Biden....      NaN   
1  democrats      It’s Time for a Blue Wave to Restore America.      NaN   
2  democrats  In Scathing Letter, More Than 80-Percent of Fa...      NaN   
3  democrats  Mark Cuban endorses Biden on Hannity: He 'actu...      NaN   
4  democrats  Biden Surge and Trump Failures Expand Electora...      NaN   

  distinguished  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [13]:
# Check for duplicates

len(democrats['title'].unique())

979

In [14]:
# Remove duplicates

democrats.drop_duplicates(subset='title', keep="first", inplace = True)

In [15]:
# Remove moderator comments as it is deemed to not be representative of organic post content

democrats = democrats[democrats['distinguished']!='moderator']

In [16]:
# Drop the "distinguised" as it no longer serves any purpose

democrats.drop('distinguished',axis=1,inplace=True)

In [17]:
# Check that column has been dropped

democrats.head()

subreddit                                              title selftext
0  democrats  Hey All! I made a documentary about Joe Biden....      NaN
1  democrats      It’s Time for a Blue Wave to Restore America.      NaN
2  democrats  In Scathing Letter, More Than 80-Percent of Fa...      NaN
3  democrats  Mark Cuban endorses Biden on Hannity: He 'actu...      NaN
4  democrats  Biden Surge and Trump Failures Expand Electora...      NaN

In [18]:
democrats.shape

(975, 3)

### Check for null values

In [19]:
# Check that column has been dropped

democrats.isnull().sum()

subreddit      0
title          0
selftext     860
dtype: int64

<font color=blue>It observed that out of 975 entries for 'selftext' column, 860 are blank. We shall let this remain and scrape the other reddit dataset for further inspection. A decision on this feature will be made after both datasets are combined.</font>

### <font color=red>Scrape Republicans Subreddit</font>

In [20]:
# Feed the subreddit URL and scrape for 20 x 50 posts (Reddit's 1,000 post limit)

web_scraper(rep_url,'21')

https://www.reddit.com/r/republican.json
12
https://www.reddit.com/r/republican.json?after=t3_hf0qsa&limit=50
27
https://www.reddit.com/r/republican.json?after=t3_hf1aiu&limit=50
9
https://www.reddit.com/r/republican.json?after=t3_he27v5&limit=50
17
https://www.reddit.com/r/republican.json?after=t3_hdsp8t&limit=50
3
https://www.reddit.com/r/republican.json?after=t3_hdhno7&limit=50
27
https://www.reddit.com/r/republican.json?after=t3_hcmpe8&limit=50
18
https://www.reddit.com/r/republican.json?after=t3_hcl5o3&limit=50
27
https://www.reddit.com/r/republican.json?after=t3_hceodb&limit=50
24
https://www.reddit.com/r/republican.json?after=t3_hboxo8&limit=50
17
https://www.reddit.com/r/republican.json?after=t3_hat4au&limit=50
4
https://www.reddit.com/r/republican.json?after=t3_h9mohc&limit=50
15
https://www.reddit.com/r/republican.json?after=t3_h9gvnb&limit=50
10
https://www.reddit.com/r/republican.json?after=t3_h85qpd&limit=50
30
https://www.reddit.com/r/republican.json?after=t3_h7tdvi&limit

In [21]:
# Load information into a dataframe and reset post for next pull

rep_data = pd.DataFrame(posts)
posts = []

In [22]:
# Check the number of posts scraped and size up data

rep_data.shape

(992, 110)

In [23]:
# Inspect the dataframe for routine check

rep_data.head()

approved_at_utc   subreddit selftext author_fullname  saved  \
0            None  Republican              t2_4xt0mqzz  False   
1            None  Republican              t2_1s4usext  False   
2            None  Republican              t2_4ywyqqu1  False   
3            None  Republican                 t2_tpsej  False   
4            None  Republican                 t2_tpsej  False   

  mod_reason_title  gilded  clicked  \
0             None       0    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  \
0  New evidence shows it is time to charge Joe Bi...                  []   
1                       Just putting this out there.                  []   
2  Welp...looks like EA and Battlefield joined th...                  []   
3  BET Founder says 'black people laugh at white ...                  []   
4  Explosive New FBI Notes Confirm Obama Directed...                  []   

  subreddit_name_prefixed  hidden  pwls link_flair_css_class  downs  \
0            r/Republican   False     6                 None      0   
1            r/Republican   False     6                 None      0   
2            r/Republican   False     6                 None      0   
3            r/Republican   False     6                 None      0   
4            r/Republican   False     6                 None      0   

   thumbnail_height top_awarded_type  hide_score       name  quarantine  \
0              93.0             None        True  t3_hflqaf       False   
1             140.0             None       False  t3_hf616r       False   
2             140.0             None        True  t3_hfmodl       False   
3              73.0             None        True  t3_hfmbjo       False   
4              97.0             None        True  t3_hfm3ky       False   

  link_flair_text_color  upvote_ratio author_flair_background_color  \
0                  dark          0.75                          None   
1                  dark          0.73                          None   
2                  dark          0.89                          None   
3                  dark          0.97                          None   
4                  dark          0.82                          None   

  subreddit_type   ups  total_awards_received media_embed  thumbnail_width  \
0         public    69                      0          {}            140.0   
1         public  2477                      0          {}            140.0   
2         public    45                      0          {}            140.0   
3         public    30                      0          {}            140.0   
4         public    18                      0          {}            140.0   

  author_flair_template_id  is_original_content user_reports secure_media  \
0                     None                False           []         None   
1                     None                False           []         None   
2                     None                False           []         None   
3                     None                False           []         None   
4                     None                False           []         None   

   is_reddit_media_domain  is_meta category secure_media_embed  \
0                   False    False     None                 {}   
1                    True    False     None                 {}   
2                    True    False     None                 {}   
3                   False    False     None                 {}   
4                   False    False     None                 {}   

  link_flair_text  can_mod_post  score approved_by author_premium  \
0            None         False     69        None          False   
1            None         False   2477        None          False   
2            None         False     45        None          False   
3        

### Save and export raw uncleaned file before treatment

In [24]:
rep_data.to_csv('../republican_uncleaned.csv', index=False, encoding='utf-8-sig')

In [25]:
# Load the uncleaned file into a new dataframe for cleaning

republicans = pd.read_csv('../republican_uncleaned.csv')

In [26]:
# The ones with discussion text for analysis are mainly title, selftext and comments. (comments excluded due to volume)
# "Distinguished" will be pulled as well to identify moderator posts and see if they should be kept

cols_to_keep = ['subreddit','title','selftext','distinguished']
republicans = republicans.loc[:, cols_to_keep]

In [27]:
# Verify that all data was carried over

republicans.shape

(992, 4)

In [28]:
republicans.head()

subreddit                                              title selftext  \
0  Republican  New evidence shows it is time to charge Joe Bi...      NaN   
1  Republican                       Just putting this out there.      NaN   
2  Republican  Welp...looks like EA and Battlefield joined th...      NaN   
3  Republican  BET Founder says 'black people laugh at white ...      NaN   
4  Republican  Explosive New FBI Notes Confirm Obama Directed...      NaN   

   distinguished  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [29]:
# Check for duplicates

len(republicans['title'].unique())

810

In [30]:
# Remove duplicates

republicans.drop_duplicates(subset='title', keep="first", inplace = True)

In [31]:
# Remove moderator comments as it is deemed to not be representative of organic post content

republicans = republicans[republicans['distinguished']!='moderator']

C:\Users\silve\anaconda3\lib\site-packages\pandas\core\ops\array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


In [32]:
# Drop the "distinguised" as it no longer serves any purpose

republicans.drop('distinguished',axis=1,inplace=True)

In [33]:
# Check that column has been dropped

republicans.head()

subreddit                                              title selftext
0  Republican  New evidence shows it is time to charge Joe Bi...      NaN
1  Republican                       Just putting this out there.      NaN
2  Republican  Welp...looks like EA and Battlefield joined th...      NaN
3  Republican  BET Founder says 'black people laugh at white ...      NaN
4  Republican  Explosive New FBI Notes Confirm Obama Directed...      NaN

In [34]:
republicans.shape

(810, 3)

### Check for null values

In [35]:
# Check that column has been dropped

republicans.isnull().sum()

subreddit      0
title          0
selftext     767
dtype: int64

<font color=blue>It observed that out of 810 entries for 'selftext' column, 767 are blank. We shall let this remain and a decision on this feature will be made after both datasets are combined.</font>

### Balance Post Counts

<font color=blue>The number of Democrat data points of 975 are much larger than the Republican's 810. In order get an even accuracy, data points will be dropped from Democrats set to match the number of the Republicans.</font>

In [36]:
# Drop the needed number of rows

democrats = democrats.iloc[:-165]

In [37]:
# Ensure that the number of democrat subreddit now matches republican

democrats.shape

(810, 3)

### Save each of the individual Subreddit  Data

In [39]:
democrats.to_csv('../democrats.csv', index=False, encoding='utf-8-sig')
republicans.to_csv('../republicans.csv', index=False, encoding='utf-8-sig')

### Merge both Subreddit Data Sets

In [40]:
# Combine both subreddit data

combined_df = pd.concat([democrats, republicans], ignore_index=True)

In [41]:
# Check that it has been correctly merged. Top half will belong to democrats subreddit

combined_df.head()

subreddit                                              title selftext
0  democrats  Hey All! I made a documentary about Joe Biden....      NaN
1  democrats      It’s Time for a Blue Wave to Restore America.      NaN
2  democrats  In Scathing Letter, More Than 80-Percent of Fa...      NaN
3  democrats  Mark Cuban endorses Biden on Hannity: He 'actu...      NaN
4  democrats  Biden Surge and Trump Failures Expand Electora...      NaN

In [42]:
# Check that it has been correctly merged. Bottom half will belong to republican subreddit

combined_df.tail()

subreddit                                              title selftext
1615  Republican  Forbes: Minneapolis Votes To Disband Police De...      NaN
1616  Republican                       Why isn't it viral yet, PT.2      NaN
1617  Republican    Pelosi Delivers Press Conference in Racist Garb      NaN
1618  Republican                 America Is In A Cultural Civil War      NaN
1619  Republican  Yahoo News: " Why it might be time to finally ...      NaN

In [43]:
# Check that datapoint collection are of correct size

combined_df.shape

(1620, 3)

### Clean the Merged Data

In [44]:
# Remove URL prefixes and numbers. The rest of special characters will be removed during modeling 

combined_df['title'] = combined_df['title'].str.replace(r'https://', '').replace(r'\d+', '')

In [45]:
# Do a routine check to ensure no abnormality after the clean

combined_df.head()

subreddit                                              title selftext
0  democrats  Hey All! I made a documentary about Joe Biden....      NaN
1  democrats      It’s Time for a Blue Wave to Restore America.      NaN
2  democrats  In Scathing Letter, More Than 80-Percent of Fa...      NaN
3  democrats  Mark Cuban endorses Biden on Hannity: He 'actu...      NaN
4  democrats  Biden Surge and Trump Failures Expand Electora...      NaN

### Evaluation of Missing Values

In [46]:
combined_df.isnull().sum()

subreddit       0
title           0
selftext     1488
dtype: int64

In [47]:
combined_df.shape

(1620, 3)

<font color=blue>From the above, we can see that only 8% of the selftext field contains text, hence a decision has been made to drop the data since there won't be much information loss. A lot of the posts contain headlines from articles. The good thing about headlines is that they usually capture the most newsworthy and salient summary of the article. Hence, the title feature alone would be deemed sufficient to carry words useful for prediction.

Therefore, the 1,620 datapoints would be deemed sufficient to generate sigficant result due to the above reason. 

</font>

In [48]:
# Drop the selftext column

combined_df.drop('selftext',axis=1,inplace=True)

### Save Merged Data

In [50]:
combined_df.to_csv('../project_data.csv', index=False, encoding='utf-8-sig')